eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java

```diff
@@ -154,48 +162,170 @@ private void getOrCreateClusteredVariantAccessions(List<? extends SubmittedVaria
      */
-    private List<IClusteredVariant> processClusteredRemappedVariantsWhereNoRSExists(List<? extends SubmittedVariantEntity> submittedVariantEntities) {
-        List<SubmittedVariantEntity> submittedVariantWithRS = submittedVariantEntities.stream()
+    private List<SubmittedVariantEntity> processClusteredRemappedVariantsRSSplit(List<? extends SubmittedVariantEntity> submittedVariantEntities)
+            throws AccessionCouldNotBeGeneratedException{
+        List<SubmittedVariantEntity> clusteredRemappedSubmittedVariants = getClusteredAndRemappedVariants(submittedVariantEntities);
+        Map<Long,List<SubmittedVariantEntity>> submittedVariantsEligibleForRSSplit= getVariantsEligibleForRSSplit(clusteredRemappedSubmittedVariants);
+        if(!submittedVariantsEligibleForRSSplit.isEmpty()){
+            List<SubmittedVariantEntity> processedVariants = processRSSplitVariants(submittedVariantsEligibleForRSSplit);
+            return processedVariants;
+        }
+        return Collections.emptyList();
+    }
+
+    private List<SubmittedVariantEntity> getClusteredAndRemappedVariants(List<? extends SubmittedVariantEntity> submittedVariants){
+        return submittedVariants.stream()
                 .filter(v -> v.getClusteredVariantAccession() != null)
                 .filter(v -> !StringUtil.isBlank(v.getRemappedFrom()))
                 .collect(Collectors.toList());
-        List<ClusteredVariantEntity> clusteredVariantEntityNoRSExists = getClusteredVariantEntityNoRSExists(submittedVariantWithRS);
-        mongoTemplate.insert(clusteredVariantEntityNoRSExists, ClusteredVariantEntity.class);
-        clusteringCounts.addClusteredVariantsRemapped(clusteredVariantEntityNoRSExists.size());
-        return clusteredVariantEntityNoRSExists.stream()
-                .map(cve -> cve.getModel())
-                .collect(Collectors.toList());
     }
 
-    private List<ClusteredVariantEntity> getClusteredVariantEntityNoRSExists(List<? extends SubmittedVariantEntity> submittedVariantWithRS) {
-        List<IClusteredVariant> clusteredVariantList = submittedVariantWithRS.stream()
+    private Map<Long,List<SubmittedVariantEntity>> getVariantsEligibleForRSSplit(List<? extends SubmittedVariantEntity> submittedVariants) {
+        List<IClusteredVariant> clusteredVariants = submittedVariants.stream()
                 .map(this::toClusteredVariantEntity)
                 .collect(Collectors.toList());
-        List<String> existingClusteredVariantsHashes = clusteredService.get(clusteredVariantList).stream()
+        List<String> clusteredVariantsRSExistsInDB = clusteredService.get(clusteredVariants).stream()
                 .map(AccessionWrapper::getHash)
                 .collect(Collectors.toList());
+        Map<Long,List<SubmittedVariantEntity>> submittedVariantRSSplit = submittedVariants.stream()
+                .filter(sve -> !clusteredVariantsRSExistsInDB.contains(toClusteredVariantEntity(sve).getHashedMessage()))
```

Comment:
@andresfsilva @nitin-ebi  From a performance standpoint, I was wondering if it would be worthwhile to make clusteredVariantsRSExistsInDB a map instead of a list given that we are traversing that multiple times for every submitted variant entry?

Reply:
Also, some illustration as to why this selection criteria works would be helpful.

Reply:
Checked and have refactored the code to use HashSet and HashMap for faster iterations wherever we need to search or filter based on hash.